File: src/util/graphics/drag/Placeholder.java
package util.graphics.drag;

import de.jensd.fx.glyphs.GlyphIcons;
import gui.objects.icon.Icon;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.layout.Pane;
import javafx.scene.layout.StackPane;
import org.reactfx.Subscription;
import static util.dev.Util.noØ;
import static util.graphics.Util.layHeaderBottom;
import static util.graphics.UtilKt.removeFromParent;
import static util.reactive.Util.maintain;

/**
 * Placeholder pane. Can invoke action and display its icon and name.<br/>
 * Useful mostly instead of content when there is none, e.g., "click to add items". Or to highlight on hover/mouse
 * over to signal possible interaction.
 */
public class Placeholder extends StackPane {

	private static final String STYLECLASS = "placeholder-pane";
	private static final String STYLECLASS_ICON = "placeholder-pane-icon";

	public final Icon icon = new Icon().styleclass(STYLECLASS_ICON);
	public final Label desc = new Label();
	private Subscription s;
//	private Pane parent;

	/**
	 * @param actionIcon non null icon
	 * @param actionName non null action name
	 * @param action non null action
	 */
	public Placeholder(GlyphIcons actionIcon, String actionName, Runnable action) {
		icon.icon(noØ(actionIcon));
		desc.setText(noØ(actionName));
		icon.onClick(noØ(action));
		getStyleClass().add(STYLECLASS);
		setOnMouseClicked(e -> { action.run(); e.consume(); });
//		setOnKeyPressed(e -> { if (ENTER) action.run(); e.consume(); });   // TODO: implement sth along these lines
		getChildren().add(layHeaderBottom(8, Pos.CENTER, icon, desc));
		setVisible(false);
	}

	/**
	 * Shows this placeholder for given node or does nothing if node null.
	 * Use {@link #visibleProperty()} to observe visibility.
	 *
	 * @param n nullable node
	 */
	public void showFor(Node n) {
		Pane p = n instanceof Pane ? (Pane) n : n==null || n.getParent()==null ? null : (Pane) n.getParent();
		if (p!=null && !p.getChildren().contains(this)) {
//            parent = p;
//            parent.getChildren().forEach(c -> c.setOpacity(0.2));
			setVisible(true);
			p.getChildren().add(this);
			s = maintain(n.layoutBoundsProperty(), b -> {
				double w = b.getWidth();
				double h = b.getHeight();
				setMaxSize(w, h);
				setPrefSize(w, h);
				setMinSize(w, h);
				resizeRelocate(b.getMinX(), b.getMinY(), w, h);
			});
			toFront();
		}
	}

	/**
	 * Use {@link #visibleProperty()} to observe visibility.
	 */
	public void hide() {
//        if (parent!=null) parent.getChildren().forEach(c -> c.setOpacity(1));
		if (s!=null) s.unsubscribe();
		removeFromParent(this);
		setVisible(false);
	}

	/**
	 * Use {@link #visibleProperty()} to observe visibility.
	 */
	public void show(Node n, boolean visible) {
		if (visible) showFor(n);
		else hide();
	}
}

File: src/player/main/sp/it/pl/gui/objects/placeholder/Placeholder.kt
package sp.it.pl.gui.objects.placeholder

import de.jensd.fx.glyphs.GlyphIcons
import javafx.geometry.Pos
import javafx.scene.Node
import javafx.scene.input.KeyCode.ENTER
import javafx.scene.input.KeyEvent.KEY_PRESSED
import javafx.scene.input.MouseButton.PRIMARY
import javafx.scene.input.MouseEvent.MOUSE_CLICKED
import javafx.scene.layout.Pane
import javafx.scene.layout.StackPane
import javafx.scene.text.TextAlignment
import sp.it.pl.gui.objects.Text
import sp.it.pl.gui.objects.icon.Icon
import sp.it.util.functional.traverse
import sp.it.util.reactive.Subscription
import sp.it.util.reactive.onEventDown
import sp.it.util.reactive.sync
import sp.it.util.ui.Util.layHeaderBottom
import sp.it.util.ui.containsMouse
import sp.it.util.ui.lay
import sp.it.util.ui.maxSize
import sp.it.util.ui.minSize
import sp.it.util.ui.prefSize
import sp.it.util.ui.pseudoClassChanged
import sp.it.util.ui.removeFromParent
import sp.it.util.ui.size

/**
 * Placeholder pane. Can invoke action and display its icon and multi-line information text.
 *
 * Useful mostly instead of empty content like "click to add items" or to signal interactive ui element.
 */
open class Placeholder(actionIcon: GlyphIcons, actionName: String, action: () -> Unit): StackPane() {

    @JvmField val icon = Icon()
    @JvmField val info = Text()
    private var s: Subscription? = null

    init {
        styleClass += STYLECLASS
        info.text = actionName
        info.textAlignment = TextAlignment.CENTER

        icon.styleclass(STYLECLASS_ICON)
        icon.icon(actionIcon)
        icon.onClick(action)
        icon.isMouseTransparent = true
        icon.isFocusTraversable = true
        hoverProperty() sync { icon.select(it) }

        onEventDown(MOUSE_CLICKED, PRIMARY) { action() }
        onEventDown(KEY_PRESSED, ENTER) { action() }
        parentProperty() sync {
            if (it!=null && it.scene?.window?.isShowing==true && it.containsMouse())
                pseudoClassChanged("hover", true)
        }

        lay += layHeaderBottom(8.0, Pos.CENTER, icon, info)
        isVisible = false
    }

    /** Shows this placeholder so it is topmost child of specified node (or its first [Pane] parent), [visible] is true and [icon].[focused] is true. */
    fun showFor(n: Node) {
        val p = n.traverse { it.parent }.filterIsInstance<Pane>().firstOrNull()
        if (p!=null && this !in p.children) {
            p.lay += this
            s?.unsubscribe()
            s = n.layoutBoundsProperty().sync {
                minSize = it.size
                prefSize = it.size
                maxSize = it.size
                resizeRelocate(it.minX, it.minY, it.width, it.height)
            }
            toFront()
            isVisible = true
            requestFocus()
        }
    }

    override fun requestFocus() {
        icon.requestFocus()
    }

    /** Hides this placeholder so it is not a child of any node and [visible] is false. */
    fun hide() {
        s?.unsubscribe()
        s = null
        removeFromParent()
        isVisible = false
    }

    /** Calls [show] if specified visibility is true and [hide] otherwise. */
    fun show(n: Node, visible: Boolean) {
        if (visible) showFor(n)
        else hide()
    }

    companion object {
        private const val STYLECLASS = "placeholder-pane"
        private const val STYLECLASS_ICON = "placeholder-pane-icon"
    }

}